FSI - Work in progress to get a working version

gantech opened the pull request:
Ready for merge if clears formatting and other checks

Commit:
Pass the openfast_fsi node to the constructor

Instead of the full Realm node

Commit:
Move bulk from constructor to setup phase for FSIturbine

Commit:
Change FSIturbine constructor to use naluOutputP0

Commit:
Declaration of div_mesh_vel

Commit:
Set meshMotion in SolutionOptions to true for fsi in aeroModels

Commit:
Rearrange mesh motion related work

FSI's displacements were originally mapped and applied by FrameOpenFAST inside
mesh_motion. However, now that can just be done inside Realm and using
set_current_coordinates.

In the process I understood this, SolutionOptions->meshMotion_ only turns on if
mesh_motion is enabled through a node in the input yaml file. I have rearranged
Realm->does_mesh_move() to also check for aeroModels_->has_fsi() and many places
like resetting EquationSystems and associated acitivities to check for
Realm->does_mesh_move() instead of SolutionOptions->meshMotion_.

I need some help to figure out if I missed any places.

Commit:
Add CalcLoads to calculate loads for FSI

Commit:
Use CalcLoads and remove bugs in FSIturbine

FSIturbine had a major bug in loadMap and loadMapInterp - They were being declared and initialized as VectorFields even though they should've been GenericFields with 4 (# SCS's) values per element face on the boundaries.

Also FSIturbine now uses CalcLoads to calculate a single total force at each SCS ip rather than separate pressure and viscous forces

Mapping displacements turned off for now until test runs

Commit:
Move aeroModels_ initialization from Realm epilog to prolog

Commit:
Map displacements after initialization

psakievich commented on src/Realm.C:
```diff
@@ -63,6 +63,7 @@
 
 // actuator line/fsi
 #include <aero/AeroContainer.h>
+#include <aero/fsi/OpenfastFSI.h>
```
```suggestion
```
This header should not be needed.  All dependencies on FSI code should be isolated to the AeroContainer

Commit:
Change blade displacements to be same as tower/nacelle/hub

Remove a whole bunch of cerr debug print statements

Commit:
Change openfast calls to use double ptrs instead of vector doubles

Remove debug cerr print statements

Uncommment all displacement and load mapping calls

Commit:
Set fsiContainer to null ptr if openfast_fsi is not present

Other cosmetic changes

Commit:
FSI changes for restart and process_multi_physics_transfer

Map displacements on restart with FSI

Reorder process_multi_physics transfer to call OpenfastFSI

gantech committed:
Update src/Realm.C

Co-authored-by: psakievich <psakiev@sandia.gov>

psakievich committed:
format

psakievich committed:
Fix some typos

psakievich commented:
I've tested this on device.  It adds several warnings from the master element functions we are using on host that are `__host__ __device` functions, but it is okay since we are only using them on host.  

Also the openfast based unit tests still segfault using the fsi branch, but we already knew that.  Something to track down still. 

We should also delete the extra print statements, but since we are going to refactor this soon and this code is all isolated unless doing FSI work I'm inclined to leave them for now.

Those seem to be the only outstanding issues and I don't think they should keep us from merging.

psakievich committed:
Merge branch 'master' into f/fsi_wip

psakievich committed:
Merge branch 'master' into f/fsi_wip

gantech commented:
Thanks Phil. @ashesh2512 Could you please go over this and see if we missed anything?

ashesh2512 commented on src/aero/fsi/FSIturbine.C:
```diff
@@ -1274,41 +1285,43 @@ fsiTurbine::computeHubForceMomentForPart(
-    }
-  }
+  // TODO: This is looping over the wrong buckets - Nodes instead of faces
+  // Is this even required anymore? Probly can delete
```
does this need to be deleted?

ashesh2512 commented on src/aero/fsi/FSIturbine.C:
```diff
@@ -1968,46 +1933,37 @@ fsiTurbine::mapDisplacements()
-          brFSIdata_.bld_pitch[iBlade],
-          &brFSIdata_.bld_root_def[iBlade * 6 + 3],
-          brFSIdata_.bld_rloc[*dispMapNode + iStart]);
+        computeDisplacement(totDispNode.data(), tmpNodePos.data(), dx, oldxyz);
```
does this need to be deleted?

gantech replied:
Agreed. Will delete

ashesh2512 commented on src/aero/fsi/FSIturbine.C:
```diff
@@ -1968,46 +1933,37 @@ fsiTurbine::mapDisplacements()
+        computeDisplacement(totDispNode.data(), tmpNodePos.data(), dx, oldxyz);
+        //       brFSIdata_.bld_pitch[iBlade],
+        //       &brFSIdata_.bld_root_def[iBlade * 6 + 3],
+        //       brFSIdata_.bld_rloc[*dispMapNode + iStart]
```
does this need to be deleted?

ashesh2512 commented on src/aero/fsi/FSIturbine.C:
```diff
@@ -1968,46 +1933,37 @@ fsiTurbine::mapDisplacements()
+    // NaluEnv::self().naluOutputP0() << "Error in displacement for blade " <<
+    // iBlade << " = " <<
+    //    errorNorm[0] << " " << errorNorm[1] << " " << errorNorm[2] <<
+    //    std::endl ;
```
does this need to be deleted?

ashesh2512 commented on src/aero/fsi/FSIturbine.C:
```diff
@@ -1968,46 +1933,37 @@ fsiTurbine::mapDisplacements()
+    // NaluEnv::self().naluOutputP0() << "Error in velocity for blade " <<
+    // iBlade << " = " <<
+    //    errorNorm[0] << " " << errorNorm[1] << " " << errorNorm[2] <<
+    //    std::endl ;
```
does this need to be deleted?

ashesh2512 commented on src/aero/fsi/FSIturbine.C:
```diff
@@ -2717,31 +2640,31 @@ fsiTurbine::computeLoadMapping()
+  //         NaluEnv::self().naluOutputP0() << "i = " << i << ", " <<
+  //         bld_rmm_[iStart+i][0] <<
+  //               " - " << bld_rmm_[iStart+i][1] << ", dr = " <<
+  //               bld_dr_[iStart+i] << std::endl ;
```
Is this left over code?

ashesh2512 commented:
Looks good to me other than a bunch of commented out code in `FSITurbine.C`. If that is intentional, I am good with this PR.

psakievich commented:
@gantech is the commented code part of your testing infrastructure or can we delete it?

gantech commented:
Thanks Ashesh.. I'm addressing your comments now

gantech commented on src/Realm.C:
```diff
@@ -3427,15 +3441,31 @@ Realm::populate_restart(double& timeStepNm1, int& timeStepCount)
+
       // reset the current time for the meshMotionAlgs
-      meshMotionAlg_->restart_reinit(foundRestartTime);
+      if (meshMotionAlg_)
```
Agreed!

gantech commented on src/Realm.C:
```diff
@@ -3427,15 +3441,31 @@ Realm::populate_restart(double& timeStepNm1, int& timeStepCount)
       compute_geometry();
-      meshMotionAlg_->post_compute_geometry();
+
+      if (meshMotionAlg_)
```
Agreed!

gantech commented on src/aero/AeroContainer.C:
```diff
@@ -44,14 +44,14 @@ AeroContainer::AeroContainer(const YAML::Node& node)
-  NaluParsingHelper::find_nodes_given_key("openfast_fsi", node, foundFsi);
-  if (foundFsi.size() > 0) {
+  // std::vector<const YAML::Node*> foundFsi;
+  // NaluParsingHelper::find_nodes_given_key("openfast_fsi", node, foundFsi);
```
Yeah.. This code should've worked. I'd like to leave it in there if we want to fix this later

Commit:
Addressing some of Ashesh's comments

psakievich commented:
@ashesh2512 I'm going to merge after this CI passes. I will continue to address these issues in #1098 .  We're going to try and merge that soon after this PR.
